scripts/core/editor3/helpers/tansa.ts

```diff
@@ -157,30 +157,124 @@ function updateText(editorState, content, block, newText, diffMatchPatch) {
     let newContent = content;
     let offset = 0;
     let diffs;
-    let selection;
+    let previousDiff;
 
     diffs = diffMatchPatch.diff_main(text, newText);
 
     diffs.forEach((diff) => {
         const _text = diff[1];
 
         if (diff[0] === 0) {
+            if (previousDiff != null) {
+                const _previousText  = previousDiff[1];
+
+                newContent = removeText(editorState, newContent, block, offset, _previousText);
+            }
             offset += _text.length;
+            previousDiff = null;
         } else if (diff[0] === 1) {
-            selection = createSelectionForBlock(editorState, block, offset);
-            offset += _text.length;
+            if (previousDiff == null) {
+                ({newContent, offset} = insertText(editorState, newContent, block, offset, _text));
+            } else {
+                const _previousText  = previousDiff[1];
 
-            newContent = Modifier.insertText(newContent, selection, _text);
+                ({newContent, offset}  =  replaceText(editorState, newContent, block, offset, _previousText, _text));
+                previousDiff = null;
+            }
         } else {
-            selection = createSelectionForBlock(editorState, block, offset, _text.length);
+            if (previousDiff != null) {
+                const _previousText  = previousDiff[1];
+
+                newContent = removeText(editorState, newContent, block, offset, _previousText);
+            }
 
-            newContent = Modifier.removeRange(newContent, selection, 'forward');
+            previousDiff = diff;
         }
     });
 
     return newContent;
 }
 
+/**
+ * Insert a new text at offset position
+ *
+ * @param {EditorState} editorState
+ * @param {ContentState} content
+ * @param {Block} block
+ * @param {Integer} offset
+ * @param {String} text
+ * @returns {ContentState, Integer}
+ */
+function insertText(editorState, content, block, offset, text) {
+    const selection = createSelectionForBlock(editorState, block, offset);
+    const newContent = Modifier.insertText(content, selection, text);
+
+    offset += text.length;
+    return {newContent, offset};
+}
+
+/**
+ * Replate text with newText and preserve the existing inline styles and entities
```

Comment:
I wouldn't try that, it's error prone imo, dropping styles and entities should be fine for now

Reply:
It should be ok since I replaced the text character by character and I preserved the inline styles and entity.

Reply:
but the words order can be different too

Reply:
For uppercase and lowercase no and probably for other cases. For translation yes.

Reply:
Maybe we can set on macro if we want to preserve or not the style and entity. On translation I think it is not possible to preserve them.

Reply:
ah right, I was talking about translation, uppercase/lowercase should be fine